tsl/test/src/remote/node_killer.c

```diff
@@ -17,88 +17,142 @@
+	{
+		elog(WARNING, "kill event: %s", remote_dist_txn_event_name(rnk->killevent));
+		remote_node_killer_kill(&rnk_event);
+		remote_dist_txn_set_event_handler(NULL);
+	}
+}
+
+static DistTransactionEventHandler eventhandler = {
+	.handler = kill_backend,
+	.data = &rnk_event,
+};
 
 TS_FUNCTION_INFO_V1(ts_remote_node_killer_set_event);
 
-RemoteNodeKiller *
-remote_node_killer_init(TSConnection *conn)
+void
+remote_node_killer_init(RemoteNodeKiller *rnk, const TSConnection *conn,
+						const DistTransactionEvent event)
 {
-	RemoteNodeKiller *rnk = palloc(sizeof(RemoteNodeKiller));
+	int remote_pid = remote_connecton_get_remote_pid(conn);
+
+	if (remote_pid == -1)
+		elog(ERROR, "could not get PID of remote backend process");
 
+	MemSet(rnk, 0, sizeof(*rnk));
 	rnk->conn = conn;
-	rnk->pid = remote_connecton_get_remote_pid(conn);
+	rnk->pid = remote_pid;
+	rnk->killevent = event;
+}
 
-	/* do not throw error here on pid = 0 to avoid recursive abort */
-	/* remote_connection_report_error(ERROR, res, conn, false, sql);  */
-	return rnk;
+static int
+kill_backends(pid_t pid, int sig)
+{
+#ifdef HAVE_SETSID
+	/* If PostgreSQL is compiled with setsid support, then the running
+	 * PostgreSQL processes are in a process group and the negative pid kills
+	 * all the other processes in the same group. */
+	return kill(-pid, SIGTERM);
+#else
+	return kill(pid, SIGTERM);
+#endif
 }
 
+/*
+ * Kill a remote process (data node).
+ *
+ * This only works if the process is on the same instance, which is true for
+ * regression tests.
+ *
+ * The function should not throw errors since it might be invoked from a
+ * transaction abort handler, which itself is invoked by errors.
+ */
 void
 remote_node_killer_kill(RemoteNodeKiller *rnk)
 {
+	PGPROC *proc;
+	int ret;
+
+	if (rnk->num_kills > 0)
+		elog(WARNING, "%s: cannot kill twice on the same event", __func__);
+
 	/*
 	 * do not use pg_terminate_backend here because that does permission
 	 * checks through the catalog which requires you to be in a transaction
 	 */
-	PGPROC *proc = BackendPidGetProc(rnk->pid);
+	proc = BackendPidGetProc(rnk->pid);
 
 	if (proc == NULL)
 		ereport(WARNING, (errmsg("PID %d is not a PostgreSQL server process", rnk->pid)));
-	kill_event = NULL;
-#ifdef HAVE_SETSID
-	if (kill(-rnk->pid, SIGTERM))
-#else
-	if (kill(rnk->pid, SIGTERM))
-#endif
+
+	rnk->num_kills++;
+
+	ret = kill_backends(rnk->pid, SIGTERM);
+
+	if (ret != 0)
 		ereport(WARNING, (errmsg("could not send signal to process %d: %m", rnk->pid)));
+	else
+	{
+		/* Ensure that the backend is dead before proceeding. Otherwise, we
+		 * might end up in a race condition where we check the status of the
+		 * backend before it has actually exited, leading to flaky testing. */
+		unsigned wait_count = 100;
+
+		while (BackendPidGetProc(rnk->pid) != NULL)
+		{
+			if (wait_count == 0)
+			{
+				elog(WARNING, "timeout while waiting for killed backend to exit");
+				break;
```

Comment:
I think you should just kill it with `SIGKILL` and be done with it. If it doesn't respond, something is very wrong.

Reply:
Not sure what you mean by "respond". Killing it means sending a signal, it doesn't imply immediate or synchronous exit of the backend. You need something like `waitpid` for that, but it is not possible because it is not portable to windows.

Polling the backend is a reasonable alternative IMO.